helix-core/src/increment/date_time.rs

```diff
@@ -0,0 +1,512 @@
+            DateUnit::AmPm => toggle_am_pm(self.date_time),
+        }
+        .unwrap_or(self.date_time);
+
+        (self.range, date_time.format(self.fmt).to_string().into())
+    }
+}
+
+static FORMATS: Lazy<Vec<Format>> = Lazy::new(|| {
+    vec![
+        Format::new("%Y-%m-%d %H:%M:%S").unwrap(), // 2021-11-24 07:12:23
+        Format::new("%Y/%m/%d %H:%M:%S").unwrap(), // 2021/11/24 07:12:23
+        Format::new("%Y-%m-%d %H:%M").unwrap(),    // 2021-11-24 07:12
+        Format::new("%Y/%m/%d %H:%M").unwrap(),    // 2021/11/24 07:12
+        Format::new("%Y-%m-%d").unwrap(),          // 2021-11-24
+        Format::new("%Y/%m/%d").unwrap(),          // 2021/11/24
+        Format::new("%a %b %d %Y").unwrap(),       // Wed Nov 24 2021
+        Format::new("%d-%b-%Y").unwrap(),          // 24-Nov-2021
+        Format::new("%Y %b %d").unwrap(),          // 2021 Nov 24
+        Format::new("%b %d, %Y").unwrap(),         // Nov 24, 2021
+        Format::new("%-I:%M:%S %P").unwrap(),      // 7:21:53 am
+        Format::new("%-I:%M %P").unwrap(),         // 7:21 am
+        Format::new("%-I:%M:%S %p").unwrap(),      // 7:21:53 AM
+        Format::new("%-I:%M %p").unwrap(),         // 7:21 AM
+        Format::new("%H:%M:%S").unwrap(),          // 23:24:23
+        Format::new("%H:%M").unwrap(),             // 23:24
+    ]
+});
+
+#[derive(Clone, Debug)]
+struct Format {
+    fmt: &'static str,
+    fields: Vec<DateField>,
+    regex: Regex,
+    max_len: usize,
+}
+
+impl Format {
+    fn new(fmt: &'static str) -> Result<Self, FormatError> {
+        let mut remaining = fmt;
+        let mut fields = Vec::new();
+        let mut regex = String::new();
+        let mut max_len = 0;
+
+        while let Some(i) = remaining.find('%') {
+            let after = &remaining[i + 1..];
+            let mut chars = after.chars();
+            let c = chars
+                .next()
+                .ok_or(FormatError::UnexpectedEndOfFormatString)?;
+
+            let spec_len = if c == '-' {
+                if let Some(c) = chars.next() {
+                    1 + c.len_utf8()
+                } else {
+                    return Err(FormatError::UnexpectedEndOfFormatString);
+                }
+            } else {
+                c.len_utf8()
+            };
+
+            if i < remaining.len() - spec_len {
+                let specifier = &after[..spec_len];
+                if let Some(field) = DateField::from_specifier(specifier) {
+                    fields.push(field);
+                    max_len += field.max_len + remaining[..i].len();
+                    regex += &remaining[..i];
+                    regex += &format!("({})", field.regex);
+                    remaining = &after[spec_len..];
+                } else {
+                    return Err(FormatError::UnsupportedSpecifier(
+                        &remaining[i..i + 1 + spec_len],
+                    ));
```

Comment:
I think it's better to panic here since the format is specified by us and it's checked in tests. Making the API more ergonomic since it probably won't happen at runtime as we checked it ourselves.

Unless users have the ability to break it then we should keep it, but I don't see that now.

Reply:
I was thinking in the future we would allow users to add their own formats. Since that's currently not possible I'll get rid of `FormatError` and panic instead.